Geometry/Surfaces/QuadSurfaces/cone_class.f90

```diff
@@ -0,0 +1,451 @@
+    real(defReal)                           :: dist
+    real(defReal), dimension(3)             :: diff, orientation
+    real(defReal)                           :: a, b, c, delta, d1, d2, near, far, &
+                                               test_near, test_far
+    logical(defBool)                        :: check
+    real(defReal), parameter :: FP_MISS_TOL = ONE + 10.0_defReal * epsilon(ONE)
+
+    ! Vector for the difference between position provided and vertex
+    diff = r - self % vertex
+
+    ! Explicitly specify cone orientation
+    orientation(self % plane) = ZERO
+    orientation(self % axis)  = sign(ONE, self % hMin)
+
+    ! Calculate quadratic components in the cone
+    a = dot_product(u(self % plane), u(self % plane)) - &
+        self % tanSquare * u(self % axis)**2
+    b = dot_product(diff(self % plane) , u(self % plane)) - &
+        self % tanSquare * diff(self % axis) * u(self % axis)
+    c = dot_product(diff(self % plane), diff(self % plane)) - &
+        self % tanSquare * diff(self % axis) ** 2
+
+    ! Calculate delta/4
+    delta = b * b - a * c
+
+    ! Calculate the distances from the cone surface
+    if (abs(a) < epsilon(ONE)) then   ! One intersection: particle direction tangent to cone opening
+      d1 = - HALF * c / b
+      d2 = sign(INF, dot_product(u, orientation))
+
+      ! Check that the intersection found is in the right hemicone. Otherwise set both to INF
+      ! The sign of inf depends on the direction of the particle and the orientation of the cone
+      check = dot_product((diff + d1 * u), orientation) > ZERO
+      if (.not. check) d1 = d2
+
+    elseif (delta < epsilon(ONE)) then        ! No intersection (should never happen)
+      d1 = -INF
+      d2 = INF
+
+    else                              ! Two intersections
+      d1 = - (b + sqrt(delta)) / a
+      d2 = - (b - sqrt(delta)) / a
+
+      ! Check that the intersections found are in the right hemicone. Otherwise set to INF
+      ! The sign of inf depends on the direction of the particle and the orientation of the cone
+      check = dot_product((diff + d1 * u), orientation) > ZERO
+      if (.not. check) d1 = sign(INF, dot_product(u, orientation))
+
+      check = dot_product((diff + d2 * u), orientation) > ZERO
+      if (.not. check) d2 = sign(INF, dot_product(u, orientation))
+
+    end if
+
+    ! Save minimum and maximum distance from cone surface
+    near = min(d1, d2)
+    far  = max(d1, d2)
+
+    ! Calculate the distances from the basis of the cone
+    if (abs(u(self % axis)) > epsilon(ONE)) then    ! Normal intersection
+
+      d1 = (self % hMax - diff(self % axis)) / u(self % axis)
+      d2 = (self % hMin - diff(self % axis)) / u(self % axis)
+
+    else                                            ! Particle parallel to axis: check location
+
+      if (diff(self % axis) > self % hMin .and. diff(self % axis) < self % hMax) then
+        ! Inside the cone: basis intersection segment is -INF:+INF
```

Comment:
Should this be distance rather than segment?
```suggestion
        ! Inside the cone: base intersection distance is -INF:+INF
```

Reply:
I think segment is ok: each couple of distances (base1 and base2, or slanted1 and slanted2) identifies 2 points that create a segment. The intersection between the two segments is what is actually internal to the cone. This is what the code is doing, and I think the docs are consistent with the cylinder and box where this approach is used.

Reply:
I would then say the segment lies between -INF:INF. It is not clear what a segment being -INF:INF means.